### ISSUE TITLE
Async calls running on separate context in OnDeactivateAsync cause UI not to refresh view

### ISSUE THREAD

**Issue by darxis** [Caliburn-Micro/Caliburn.Micro#906]
Caliburn.Micro NuGet package version 4.0.212.

I have found a bug in the screen Activation/Deactivation logic.

The problem is when a long-running task is run within `OnDeactivateAsync` that is not run on the captured context.

Please see the reproduction example I pushed to my fork - [https://github.com/darxis/Caliburn.Micro/tree/bug-async-calls-in-ondeactivateasync](https://github.com/darxis/Caliburn.Micro/tree/bug-async-calls-in-ondeactivateasync)

Please see commit #[f136656](https://github.com/darxis/Caliburn.Micro/commit/f136656ed173960f424ad3d306b31952024acba9)

I modified the `Caliburn.Micro.Tutorial.Wpf` sample code - I modified the `ShellViewModel` to display a `TabControl` for two ViewModels - one of them is simulating a long-running task not on the captured context - it's just an `await Task.Delay(...)`:
```c#
protected override async Task OnDeactivateAsync(bool close, CancellationToken cancellationToken)
{
  // Here we simulate some long-running task that runs on another context (it also could be for example running any long-running task using Task.Run)
  // This causes the bug to trigger - the view for 'SomeOtherViewMode' doesn't show when selecting the 'Some other view' tab
  // If the following line is removed everything works fine.
  await Task.Delay(1000, cancellationToken);
  // Or any of the following:
  // await Task.Run(async () => await SomeLongRunningTaskAsync(cancellationToken), cancellationToken);
  // await Task.Run(() => Thread.Sleep(1000), cancellationToken);

  await base.OnDeactivateAsync(close, cancellationToken);
}
```

Reproduction steps:
1. Run the application
2. Click on the 'Some other view' tab
3. The tab view is not displayed, but it should be

I think it what happens is the following:
1. When you click the 'Some other view' tab the Dispatcher (or some other thread) sets the `ActiveItem` with the `SomeOtherViewModel` instance. 
2. The `ActiveItem` setter call returns before the `OnDeactivateAsync` method is fully executed as the `ActivateItemAsync` call is not awaited so the `_activeItem` is not changed because it is updated after `OnDeactivateAsync` is completed. That means that the `ActiveItem` setter call returns but the `ActiveItem` would return the old viewmodel until the `OnDeactivateAsync` completes and `_activeItem` is changed.
3. The thread/code that set the `ActiveItem` marks that property as "Dirty" and triggers refreshing of that property before the underneath `_activeItem` field is changed - you can check that by putting a breakpoint on the `ActiveItem` setter.

```c#
public T ActiveItem
{
	get
	{
		return _activeItem;
	}
	set
	{
		ActivateItemAsync(value, CancellationToken.None);
	}
}
```

```c#
protected virtual async Task ChangeActiveItemAsync(T newItem, bool closePrevious, CancellationToken cancellationToken)
{
	await ScreenExtensions.TryDeactivateAsync(_activeItem, closePrevious, cancellationToken);
	newItem = EnsureItem(newItem);
	_activeItem = newItem;
	NotifyOfPropertyChange("ActiveItem");
	if (IsActive)
	{
		await ScreenExtensions.TryActivateAsync(newItem, cancellationToken);
	}
	OnActivationProcessed(_activeItem, success: true);
}
```

Edit: I think it is a bad idea to fire and forget the call to `ActivateItemAsync(value, CancellationToken.None);` in the `ActiveItem` setter as the returned task would be unobserved. This also causes `UnobservedTaskException` events to raise when the `ActivateItemAsync` call throws an exception (for example within `OnActivateAsync` or `OnDeactivateAsync` methods).

**Comment by darxis** [comment]
@vb2ae Do you have any idea why the `NotifyOfPropertyChange("ActiveItem");` call doesn't force the UI to refresh the view after the long running task in `OnDeactivateAsync` is completed and `_activeItem` has been set?

**Comment by vb2ae** [comment]
Guessing the NotifyOfPropertyChange("ActiveItem") is not running on the ui thread.  Maybe this will work

       Dispatcher.CurrentDispatcher.BeginInvoke(NotifyOfPropertyChange("ActiveItem"));

**Comment by darxis** [comment]
> Guessing the NotifyOfPropertyChange("ActiveItem") is not running on the ui thread. Maybe this will work
> 
> ```
>    Dispatcher.CurrentDispatcher.BeginInvoke(NotifyOfPropertyChange("ActiveItem"));
> ```

Internally `NotifyOfPropertyChange` would call `OnUIThread(() => OnPropertyChanged(new PropertyChangedEventArgs("ActiveItem")));` so it must be something else

**Comment by darxis** [comment]
Okay, I managed to find out why the `TabControl` doesn't update the UI despite `PropertyChanged(ActiveItem)` being called. It is because when you click the second tab on the UI to show it then it is the `TabControl` that sets the `ActiveItem` and internally `TabControl` will remember this newly selected tab as the active tab. And later when the `PropertyChanged(ActiveItem)` is called the `TabControl` compares if the active tab it remembered is different than the new one because if that's the case there is no need to do anything. In this case it is the same tab that's why the UI doesn't update to the newly selected tab view.

So `TabControl` has the correct tab remembered but it displays the view for the old tab as `ActiveItem/_activeItem` is not updated yet because it is updated asynchronously later.

Also this bug doesn't occur if you activate the other tab programatically by calling `ActiveItem = _someOtherViewModel` from within the `Conductor`, because TabControl doesn't know the new tab yet, it get to know it only when the `NotifyOfPropertyChange(nameof(ActiveItem))` is called and then it refreshes the UI with the correct view.

TL;DR - everything is fine with the `NotifyOfPropertyChange(nameof(ActiveItem));` call and `ActiveItem` should set its backed field `_activeItem` synchronously before exiting the setter, not asynchronously within a new `Task` - the PR I submitted fixes this behavior.
